core/arch/arm/kernel/rpc_io_i2c.c

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2020 Foundries Ltd <jorge@foundries.io>
+ *
+ */
+#include <kernel/rpc_io_i2c.h>
+#include <kernel/thread.h>
+#include <mm/mobj.h>
+#include <optee_rpc_cmd.h>
+#include <string.h>
+
+#define I2C_BUFFER_LENGTH 512 /* enough for most use cases */
+static struct mobj *mobj;
+
+TEE_Result rpc_io_i2c_transfer(struct rpc_i2c_request *req, size_t *bytes)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct thread_param p[3] = { };
+
+	if (!req)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (req->buffer_len > I2C_BUFFER_LENGTH) {
+		EMSG("buffer too large to transfer");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (!mobj) {
+		mobj = thread_rpc_alloc_kernel_payload(I2C_BUFFER_LENGTH);
```

Comment:
How is this freed? For instance when switching driver from U-Boot to Linux.

Reply:
it is not freed. once the buffer is allocated to Linux is left there for any i2c comms that might be required. The performance price to allocate/deallocate on every transfer would be too expensive for some i2c devices. I was hoping this would not be an issue for a 512 bytes allocation.

Reply:
@jenswi-linaro maybe we should kick a timer to free the memory if no transfers occur in a period?

Reply:
That will not take care of the transition from U-Boot to Linux kernel. This doesn't seem to take concurrency into account. At:
https://github.com/OP-TEE/optee_os/blob/35e770df71a8c35d6c3006e25939d936dca02bbc/core/arch/arm/include/kernel/thread.h#L235-L237
There's a way of caching a shared memory allocation for the current invoke, per thread. This could perhaps be extended to be able so handle this kind of shared memory allocation too.

Reply:
I dont understand your point: what is your concern?  why do you mention the transition from uboot to linux as an issue? 

before linux boots, the user (uboot or spl) must **not** call this "service" and use the _imx_i2c_ driver instead ; once linux is up, the user  can safely use this service -which will allocate the required buffer- and must stop using the imx_i2c driver..

Reply:
> I dont understand your point: what is your concern? why do you mention the transition from uboot to linux as an issue?

I'd like to have some sensible behaviour when that happens.

> 
> before linux boots, the user (uboot or spl) must **not** call this "service" and use the _imx_i2c_ driver instead ; once linux is up, the user can safely use this service -which will allocate the required buffer- and must stop using the imx_i2c driver..

How is that enforced?

Reply:
it is not enforced; in fact I dont even understand the question: before using any driver, the user must understand what it does and how it fits in the system. is like asking how do you enforce that  when I unmap register memory, my system will not crash when later on when try to access the driver. Each driver serves a purpose and it is up to the client to know how to schedule them for their use case; some platforms might have only one chip in the i2c bus in which case they might chose not to use a trampoline and stick with the native driver. some platforms might not need i2c access before the normal world is up in which case there is no reason for them to use the native driver...and so forth.

Reply:
Considering a boot scenario with:
> (secure) early-boot -> (secure) OP-TEE -> (non-secure) U-Boot -> (non-secure) Linux

OP-TEE could very well use the I2C trampoline  when U-Boot has booted. This of course expects an OP-TEE I2C RPC driver is implemented in U-Boot, as you did for the Linux kernel.

Reply:
ah I am with you know -and thanks for clarifying- I did fail to establish the link since I didn't submit U-boot support.

So if we are discussing how things might evolve for a future user wanting to integrate with U-Boot - or some other form of pipeline- it would not be too hard to build some form of state machine for those users wanting synchronization between different i2c scenarios. Doing that was beyond the scope of this PR really but who knows, maybe the need arises sooner rather than later.

Reply:
This PR is about an interface exposed to normal world. It's not enough to just look at Linux then, we must have the big picture from the start.

Reply:
honestly I dont have a clue what you are asking. let me know if I should abandon the PR, this is getting very time consuming for something this trivial.

Reply:
Adding a new interface between secure and normal world is not a trivial task. Since we're going to be stuck with it once it's there we'd like to get it right the first time.